xtensa: make it work with DTLB misses during interrupt handling

dcpleung opened the pull request:
If there is a data TLB misses during interrupt handling, the double exception vector will be triggered for the miss and the EXCCAUSE overwritten. The DTLB miss will be handled and execution returned to the original vector code. Because of this, the EXCCAUSE read in the C handler thus is not the one that triggered the original vector code (for example, level-1 interrupt). So stash the EXCCAUSE and use it in this scenario such that we can process it correctly.

Note that the moving of assembly in the vector code is due to space constraint in the vector area. Or else it would be too large to fit there.

andyross reviewed (commented):
Hm, that's a good catch.   Wouldn't it make more sense to handle this on the other side, though?  Cache the "original" EXCCAUSE in the level 1 handler in assembly before entry to the handler (while running vector code in pinned a TLB page) and stash it in a SR (or just in the CPU struct, gen_zsr.py is going to start running out of registers soon).

Basically, as I'm understanding it: the value in EXCCAUSE can't ever be trusted in general C code because it's subject to being clobbered at any given moment by a TLB refill.  And that can still happen here, in the instructions before this fixup.  So close the gap before entry to general C code?

andyross commented:
Also, would be good to have a test case for this.  Might be as simple as "flush the TLB and irq_offload()".  If EXCCAUSE gets clobbered by the resulting entry code then the offload callback would presumably never happen.  I'm specifically thinking about this in the context of call0, which will want regression tests to look at the same kinds of failures.

dcpleung commented:
> Hm, that's a good catch. Wouldn't it make more sense to handle this on the other side, though? Cache the "original" EXCCAUSE in the level 1 handler in assembly before entry to the handler (while running vector code in pinned a TLB page) and stash it in a SR (or just in the CPU struct, gen_zsr.py is going to start running out of registers soon).
> 
> Basically, as I'm understanding it: the value in EXCCAUSE can't ever be trusted in general C code because it's subject to being clobbered at any given moment by a TLB refill. And that can still happen here, in the instructions before this fixup. So close the gap before entry to general C code?

The EXCCAUSE is stashed at beginning of the user exception vector (aka. `_Level1RealVector`). It is not the earliest due to space constraint, but should be at a point where there is no memory access (and thus won't generate TLB misses).

I think you see the checking of `is_dblexc` and assumed it is being handled as a double exception. Actually, `is_dblexc` being true is a side effect of handling DTLB miss in the double exception vector as DEPC is not cleared in that case (while other causes would go through the C handler and DEPC cleared before the end). So going in the C handler, as DEPC is not zero and it assumes it is dealing with a double exception. Though it turns out to be a good way to catch this exact scenario: DEPC != 0 and (fake) EXCCAUSE == DTLB miss.

dcpleung commented:
I will try to create a test for this.

andyross commented on arch/xtensa/core/vector_handlers.c:
```diff
@@ -229,6 +229,21 @@ void *xtensa_excint1_c(int *interrupted_stack)
 
 	is_dblexc = (depc != 0U);
+
+	if (is_dblexc && (cause == EXCCAUSE_DTLB_MISS)) {
```
OK, I'm starting to get this.  But walk me through why you need this if() expression?  The cause is in the ZSR register always, no matter whether there was a double exception or not.  We can just fetch it unconditionally and avoid the extra instructions (including the broken one to fetch the essentially useless value above!).  Might be need for a little wrapper API so that MMU=n builds can skip the extra ZSR.

andyross commented on arch/xtensa/core/vector_handlers.c:
```diff
@@ -229,6 +229,21 @@ void *xtensa_excint1_c(int *interrupted_stack)
+	if (is_dblexc && (cause == EXCCAUSE_DTLB_MISS)) {
+		/* If there is a DTLB miss during interrupt handling,
+		 * the double exception vector will be triggered for
+		 * TLB miss and EXCCAUSE overwritten. The DTLB will be
```
Also I think this is still unreliable?  EXCM is unmasked here:  We can have trapped a TLB exception any time **after** the rsr.depc above (for example due to a load the compiler added to fill something from the stack, or an iTLB fault because we crossed a page boundary in the code here), and so is_dblc is only "mostly always" right.

andyross reviewed (commented):
Ah, OK.  So this does basically what I suggested already.  One note on a theoretical hole I see, which I'm pretty sure can be fixed by just making the ZSR usage unconditional.

dcpleung committed:
xtensa: make it work with DTLB misses during interrupt handling

If there is a data TLB misses during interrupt handling,
the double exception vector will be triggered for the miss
and the EXCCAUSE overwritten. The DTLB miss will be handled
and execution returned to the original vector code. Because of
this, the EXCCAUSE read in the C handler thus is not the one
that triggered the original vector code (for example, level-1
interrupt). So stash the EXCCAUSE and use it in this scenario
such that we can process it correctly.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
I have modified the flow a bit so that EXCCAUSE_DTLB_MISS would never be propagated to the C handler under normal circumstances. If we see that in C handler, it means we are hitting this scenario. After unmasking EXCM, any TLB misses would be handled in the early vector code and will never be saved in ZSR_EXCCAUSE_SAVE.

andyross reviewed (dismissed):
OK, I'm convinced this ordering is safe.  You're only ever testing DEPC after EXCCAUSE, so if EXCCAUSE got clobbered it will show the incorrect (TLB fault) cause and thus pass the test in the if().

No reason not to +1.  But I still point out this is needlessly complicated.  What you want to be doing is testing vs. TLB faults in the (safe) entry path, and putting the correct cause into the ZSR always.  Then you use that ZSR always when MMU=y, and never touch EXCCAUSE, which is unsafe.   Less code, more obviously correct.

dcpleung commented:
Further tweaking it to avoid using another scratch register. And also stash DEPC for use in C handler (similar to stashing EXCCAUSE). This version should be a lot simpler, but might be a bit more confusing.

dcpleung commented:
Tighten the stashing code a bit by using XSR.

dcpleung commented:
This is actually not really enough, and I am working on another set of patches. So closing this one.
